server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Keywords.scala

```diff
@@ -0,0 +1,151 @@
+}
+
+case class Keywords(keywords: Set[Keyword]) {
+
+  def combiner(firstBuilder: FlagsBuilder, secondBuilder: FlagsBuilder): FlagsBuilder = firstBuilder.add(secondBuilder.build)
+  def accumulator(accumulator: FlagsBuilder, keyword: Keyword): FlagsBuilder = accumulator.add(keyword.asFlags())
+
+  /**
+   * This Keywords factory will filter out invalid keywords (containing inavlid character, being too long or being non
+   * exposed IMAP flags)
+   */
+  def strictFactory: KeywordsFactory = STRICT_KEYWORDS_FACTORY
+
+  /**
+   * This Keywords factory will throw upon invalid keywords (containing inavlid character, being too long or being non
+   * exposed IMAP flags)
+   *
+   * @throws IllegalArgumentException
+   */
+  def lenientFactory: KeywordsFactory = LENIENT_KEYWORDS_FACTORY
+
+  def asFlags(): Flags = keywords.asJava.stream().reduce(FlagsBuilder.builder, this.accumulator, this.combiner).build
+
+  def asFlagsWithRecentAndDeletedFrom(originFlags: Flags): Flags = {
+    val flags = asFlags
+    if (originFlags.contains(Flags.Flag.DELETED)) flags.add(Flags.Flag.DELETED)
+    if (originFlags.contains(Flags.Flag.RECENT)) flags.add(Flags.Flag.RECENT)
+    flags
+  }
+
+  def asMap: ImmutableMap[String, Boolean] = keywords.asJava.stream()
+    .collect(Guavate.toImmutableMap((keyword: Keyword) => keyword.getFlagName(), (_: Keyword) => Keyword.FLAG_VALUE))
+
+  def getKeywords: Set[Keyword] = keywords
+
+  def contains(keyword: Keyword): Boolean = keywords.contains(keyword)
+
+  override def equals(other: Any): Boolean =  other match {
+      case otherKeyword: Keywords => Objects.equal(keywords, otherKeyword.keywords)
+      case _ => false
+  }
+
+  override def hashCode: Int = Objects.hashCode(keywords)
+
+  override def toString: String = MoreObjects.toStringHelper(this).add("keywords", keywords).toString
+}
+
+class KeywordsFactory(val validator: KeywordsValidator, val filter: Keyword => Boolean, val toKeyword: ToKeyword) {
+    def fromSet(setKeywords: Set[Keyword]): Keywords = {
+      validator.validate(setKeywords)
+      Keywords(setKeywords.filter(filter))
+    }
+
+    def from(keywords: Keyword*): Keywords = {
+      fromSet(keywords.toSet)
+    }
+
+    def fromCollection(keywords: Set[String]): Keywords = fromSet(keywords.flatMap(toKeyword.toKeyword))
+
+    private[model] def fromMap(mapKeywords: Map[String, Boolean]): Keywords = {
+      if (mapKeywords.values.toSet.exists(value => value)) {
+          throw new IllegalArgumentException("Keyword must be true")
+      }
+      fromCollection(mapKeywords.keySet)
+    }
+
+    def fromFlags(flags: Flags): Keywords = {
+      val fromUserFlags: Set[Option[Keyword]] = flags.getUserFlags.map[Option[Keyword]](toKeyword.toKeyword).toSet
+      val fromSystemFlags: Set[Option[Keyword]] = flags.getSystemFlags.map[Option[Keyword]](Keyword.fromFlag).toSet
+
+      fromSet((fromUserFlags ++ fromSystemFlags).filter(option => option.isDefined).map(_.get))
```

Comment:
Indent

Reply:
replace '.filter(option => option.isDefined).map(_.get))'   with     '.flatten'

Reply:
> replace '.filter(option => option.isDefined).map(_.get))' with '.flatten'

How can i do that?